### PR TITLE
Added a ref Resolver

### LINKED ISSUE #104: Target-Snowflake PipelineWise Variant KeyError: 'type'
Using Tap-Clickup with the target-snowflake pipelinewise variant loader throws a KeyError on load.

Traceback goes to the `adjust_timestamps_in_record` function in stream_utils.py in the target-snowflake library.

Offending Record:
```
{"type": "RECORD", "stream": "folder_customfield", "record": {"id": "f510d927-7fab-4622-96d1-07193fc577cc", "name": "Focus Area", "type": "drop_down", "type_config": {"default": 0, "placeholder": null, "new_drop_down": true, "options": [{"id": "97a53bcc-3fa7-4160-b59a-267a1d7d9cd1", "name": "Package Analytics", "color": "#800000", "orderindex": 0}, {"id": "6fc0e3b2-5a38-405a-88cf-4676252ada82", "name": "Custom Package Analytics", "color": "#bf55ec", "orderindex": 1}, {"id": "adf3f2b0-707a-4dde-bbfa-8aa3fd9516d2", "name": "Internal Analytics", "color": "#0231E8", "orderindex": 2}, {"id": "6f7c7717-6f5c-4f02-b788-bbace80da55c", "name": "Implementation Support", "color": "#AF7E2E", "orderindex": 3}, {"id": "c1bb571b-f66a-4ea5-9ec0-2d57f0bb4f33", "name": "Workato", "color": "#e50000", "orderindex": 4}, {"id": "a0e778a2-3b71-4ff2-8d9c-10917d53edc0", "name": "Demo Development", "color": "#f9d900", "orderindex": 5}]}, "date_created": "1628720868248", "hide_from_guests": false, "required": false}, "time_extracted": "2022-01-04T17:30:44.178089Z"} 
```

Traceback:
```
2022-01-04T17:30:44.746439Z [info     ] Traceback (most recent call last): cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.748067Z [info     ]   File "/workspaces/hs-meltano/.meltano/loaders/target-snowflake/venv/bin/target-snowflake", line 8, in <module> cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.750433Z [info     ]     sys.exit(main())           cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.753501Z [info     ]   File "/workspaces/hs-meltano/.meltano/loaders/target-snowflake/venv/lib/python3.7/site-packages/target_snowflake/__init__.py", line 520, in main cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.756279Z [info     ]     persist_lines(config, singer_messages, table_cache, file_format_type) cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.759311Z [info     ]   File "/workspaces/hs-meltano/.meltano/loaders/target-snowflake/venv/lib/python3.7/site-packages/target_snowflake/__init__.py", line 143, in persist_lines cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.761874Z [info     ]     stream_utils.adjust_timestamps_in_record(o['record'], schemas[stream]) cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.766532Z [info     ]   File "/workspaces/hs-meltano/.meltano/loaders/target-snowflake/venv/lib/python3.7/site-packages/target_snowflake/stream_utils.py", line 69, in adjust_timestamps_in_record cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.769285Z [info     ]     if 'string' in schema['properties'][key]['type'] and \ cmd_type=loader job_id=clickup-to-snowflake-dev name=target-snowflake run_id=4f348796-2671-4a08-bcf7-0bdfce839b4a stdio=stderr
2022-01-04T17:30:44.773071Z [info     ] KeyError: 'type'  
```

### PR DESCRIPTION
**Pull request by visch**
Closes #104

### CODE REVIEW

File: tap_clickup/schemas/folder.json [hunk 1]
```diff
@@ -10,18 +10,50 @@
...
+                        "type": ["null", "integer", "string"]
                     },
                     "color": {
                         "type": ["null", "string"]
                     },
                     "type": {
                         "type": ["null", "string"]
                     }
+            },
+        "feature": {
```
Comment (visch) [inline tap_clickup/schemas/folder.json:22]: This doesn't look to be needed

File: tap_clickup/schemas/folder.json [hunk 2]
```diff
@@ -51,21 +83,26 @@
                         "type": ["null", "string"]
                     },
                     "space": {
-                        "type": ["null", "object"],
```
Comment (visch) [inline tap_clickup/schemas/folder.json:54]: Would rather not risk the change here, should add null back

File: meltano.yml [hunk 3]
```diff
@@ -13,6 +13,7 @@ plugins:
     - name: api_token
       kind: password
     select:
+    - '!shared_hierarchy.*'
```
Comment (visch) [inline meltano.yml:16]: Not needed

File: tap_clickup/schemas/space.json [hunk 4]
```diff
@@ -1,61 +1,88 @@
...
+                    "type": ["object", "null"],
+                    "properties": {
+                        "id": {
+                            "type": "integer"
+                        },
+                        "username": {
+                            "type": "string"
                         },
-                        "invited_by": {
-                            "$ref": "#/definitions/user"
```
Comment (visch) [inline tap_clickup/schemas/space.json:51]: Would be nice to keep this instead of duplicating what a user is

File: tap_clickup/schemas/team.json [hunk 5]
```diff
@@ -1,58 +1,123 @@
...
-                        "type": ["null", "string"]
-                    },
-                    "date_joined": {
-                        "type": ["null", "string"]
-                    },
-                    "date_invited": {
-                        "type": ["null", "string"]
-                    }
+                "id": {
+                    "type": "integer"
```
Comment (visch) [inline tap_clickup/schemas/team.json:7]: To be safe I think we should add null here, as these refs haven't been heavily tested by users.

File: tap_clickup/schemas/team.json [hunk 6]
```diff
@@ -1,58 +1,123 @@
...
-                {
+            "type": "object",
+            "properties": {
+                "user": {
                     "type": "object",
                     "properties": {
-                        "user": {
-                            "$ref": "#/definitions/user"
+                        "id": {
+                            "type": "integer"
```
Comment (visch) [inline tap_clickup/schemas/team.json:48]: Add nulls (similar to above comment)

File: tap_clickup/schemas/team.json [hunk 7]
```diff
@@ -1,58 +1,123 @@
...
                         }
                     }
-                } 
-            ]
+                },
+                "invited_by": {
+                    "type": "object",
+                    "properties": {
+                        "id": {
+                            "type": "integer"
```
Comment (visch) [inline tap_clickup/schemas/team.json:86]: Add nulls (similar to above comment)

**Comment by visch** [comment]
Worried about potential json schema failures as we haven't been able to fully test this with Data in production setups (Tested with my test instance and one live instances data, but history shows that's not enough)

**Comment by visch** [comment]
- [ ] Check for any items in def's that don't' have a null option, add null as an option

**Review by visch — COMMENTED** [review]
Need to fix those issues

**Comment by saeedzareian** [comment]
Hi, thank you for the great work and support here. 


I am using list (folder-list) and space heavily and this fix will help us a lot. Also, we found another issue that I will share shortly with you after verification. Could you please move forward with this pull request earlier?

File: tap_clickup/schemas/team.json [hunk 8]
```diff
@@ -1,58 +1,123 @@
...
+                        "date_invited": {
+                            "type": ["null", "string"]
+                        }
+                    }
+                },
+                "invited_by": {
+                    "type": ["null", "object"],
                     "properties": {
-                        "user": {
-                            "$ref": "#/definitions/user"
```
Comment (visch) [inline tap_clickup/schemas/team.json:48]: This should be added back to the json file, shouldn't have been removed as we now have a ref resolver.

**Comment by visch** [comment]
> Hi, thank you for the great work and support here.
> 
> I am using list (folder-list) and space heavily and this fix will help us a lot. Also, we found another issue that I will share shortly with you after verifying my solution. Could you please move forward with this pull request earlier?

Yeah I want to merge this, I just need to test that it works with target-postgres, and target-stitch. Issue is that I don't have functional testing setup to automate this for me (and I'd probably want to smoke test myself locally anyway) 

Next steps are

1. Verify data before and after this MR are the same in target-postgres (ie select * from old_tap_clickup except select * from new_tap_clickup and visaversa should return 0 records) , looks like `CHECKSUM TABLE old_tap_clickup , new_tap_clickup ;` would work too, if there are differences we should look at them and see if they make sense or not. 
2. Run tap-postgres target-stitch to be sure data gets through so I don't break anyone later
3. Run tap-postgres target-jsonl to be sure schema checks pass

I won't have time to do this for a bit. I''ll try to, or if someone could do these tests for me or a portion of them it'd be mightly helpful!

File: tap_clickup/schemas/space.json [hunk 9]
```diff
@@ -1,81 +1,108 @@
...
-            "type": ["array", "null"],
-            "items": [
-                {
+            "type": ["null", "object"],
+            "properties": {
+                "user": {
                     "type": "object",
                     "properties": {
-                        "user": {
-                            "$ref": "#/definitions/user"
```
Comment (visch) [inline tap_clickup/schemas/space.json:48]: Would be nice to keep this instead of duplicating what a user is

File: tap_clickup/schemas/team.json [hunk 10]
```diff
@@ -1,58 +1,123 @@
...
+                            "type": ["null", "string"]
+                        },
+                        "color": {
+                            "type": ["null", "string"]
+                        },
+                        "profilePicture": {
+                            "type": ["null", "string"]
                         },
-                        "invited_by": {
-                            "$ref": "#/definitions/user"
```
Comment (visch) [inline tap_clickup/schemas/team.json:51]: Would be nice to keep this instead of duplicating what a user is

File: tap_clickup/schemas/team.json [hunk 11]
```diff
@@ -1,58 +1,123 @@
...
+                "date_invited": {
+                    "type": ["null", "string"]
+                }
             }
         },
         "members": {
-            "type": "array",
-            "items": [
-                {
-                    "type": "object",
```
Comment (visch) [inline tap_clickup/schemas/team.json:45]: Would be nice to keep this instead of duplicating what a user is

**Comment by visch** [comment]
> Hi, thank you for the great work and support here.
> 
> I am using list (folder-list) and space heavily and this fix will help us a lot. Also, we found another issue that I will share shortly with you after verifying my solution. Could you please move forward with this pull request earlier?

There's also some duplicate in the PR, I'd like to remove some of them and replace them with Refs, if someone could to that it would be very helpful as well (I  hope it can work without changing the ref resolver we're using, if not we don't need to dive too hard we can just leave the duplication)

**Comment by saeedzareian** [comment]
> I won't have time to do this for a bit. I''ll try to, or if someone could do these tests for me or a portion of them it'd be mightly helpful!

I will ask my friends if they have capacity to help, their sprint starts the next Monday (No tomorrow).

File: tap_clickup/schemas/space.json (hunk 9 above)
Comment (visch) [reply tap_clickup/schemas/space.json:48]: Nested definitions were giving me some issues, posted https://github.com/transferwise/pipelinewise-singer-python/issues/55 instead . Created a tracking issue on our end here https://github.com/AutoIDM/tap-clickup/issues/109

File: tap_clickup/schemas/space.json (hunk 4 above)
Comment (visch) [reply tap_clickup/schemas/space.json:51]: noted in above comment

File: tap_clickup/schemas/team.json (hunk 10 above)
Comment (visch) [reply tap_clickup/schemas/team.json:51]: noted in above comment

**Comment by visch** [comment]
Data looks good from a run on my end 

> * [x]  Check for any items in def's that don't' have a null option, add null as an option

After review there were only a handful of properties without nulls, this shouldn't' hold up the merge as this PR puts us in a better place.

**Comment by visch** [comment]
> > I won't have time to do this for a bit. I''ll try to, or if someone could do these tests for me or a portion of them it'd be mightly helpful!
> 
> I will ask my friends if they have capacity to help, their sprint starts the next Monday (No tomorrow).

We're all merged in, pull the latest from pypi and you should be good to go!

**Pull request merged**